### ISSUE TITLE
java.lang.StackOverflowError during "fast" scrolling in OutlineView of a Netbeans Platform application

### ISSUE THREAD

**Issue by MKuettner** [apache/netbeans#6835]
### Apache NetBeans version

Apache NetBeans 20 release candidate

### What happened

In my Netbeans Platform based application there is a special dialog to select some kind of modules and sub-modules.
This dialog uses an OutlineView with one column and a CheckRenderDataProvider to display/manage selection.
The dialog looks like this:
![grafik](https://github.com/apache/netbeans/assets/72084885/1921a39c-2e74-46d3-bb9b-fd875493dc86)

If there are "many" nodes contained and if user scrolles really "fast" a java.lang.StackOverflowError occurs:

```
java.lang.StackOverflowError
	at org.netbeans.swing.etable.ETableColumn$ETableColumnHeaderRendererDelegate.getTableCellRendererComponent(ETableColumn.java:464)
	at org.openide.explorer.view.OutlineView$OutlineViewOutline$OutlineViewOutlineColumn$OutlineViewOutlineHeaderRenderer.getTableCellRendererComponent(OutlineView.java:2258)
	at org.netbeans.swing.etable.ETableHeader$ETableHeaderRenderer.getTableCellRendererComponent(ETableHeader.java:129)
	at org.netbeans.swing.etable.ETableColumn$ETableColumnHeaderRendererDelegate.getTableCellRendererComponent(ETableColumn.java:464)
	at org.openide.explorer.view.OutlineView$OutlineViewOutline$OutlineViewOutlineColumn$OutlineViewOutlineHeaderRenderer.getTableCellRendererComponent(OutlineView.java:2258)
	at org.netbeans.swing.etable.ETableHeader$ETableHeaderRenderer.getTableCellRendererComponent(ETableHeader.java:129)
	at org.netbeans.swing.etable.ETableColumn$ETableColumnHeaderRendererDelegate.getTableCellRendererComponent(ETableColumn.java:464)
	at org.openide.explorer.view.OutlineView$OutlineViewOutline$OutlineViewOutlineColumn$OutlineViewOutlineHeaderRenderer.getTableCellRendererComponent(OutlineView.java:2258)
	at org.netbeans.swing.etable.ETableHeader$ETableHeaderRenderer.getTableCellRendererComponent(ETableHeader.java:129)
	at org.netbeans.swing.etable.ETableColumn$ETableColumnHeaderRendererDelegate.getTableCellRendererComponent(ETableColumn.java:464)
	at org.openide.explorer.view.OutlineView$OutlineViewOutline$OutlineViewOutlineColumn$OutlineViewOutlineHeaderRenderer.getTableCellRendererComponent(OutlineView.java:2258)
	at org.netbeans.swing.etable.ETableHeader$ETableHeaderRenderer.getTableCellRendererComponent(ETableHeader.java:129)
	at org.netbeans.swing.etable.ETableColumn$ETableColumnHeaderRendererDelegate.getTableCellRendererComponent(ETableColumn.

```

It seems that this behavior has been introduced in Netbeans 18. All versions below do not show this error.

### How to reproduce

Currently i have no idea how to reproduce this issue outside of my application.

### Did this work correctly in an earlier version?

Apache NetBeans 17

### Operating System

Unix

### JDK

JDK 8, 17, 21

### Apache NetBeans packaging

Apache NetBeans platform

### Are you willing to submit a pull request?

Yes

**Comment by mbien** [comment]
> It seems that this behavior has been introduced in Netbeans 18. All versions below do not show this error.

hmm. The loop is actually quite tight:
```
	at org.netbeans.swing.etable.ETableColumn$ETableColumnHeaderRendererDelegate.getTableCellRendererComponent(ETableColumn.java:464)
	at org.openide.explorer.view.OutlineView$OutlineViewOutline$OutlineViewOutlineColumn$OutlineViewOutlineHeaderRenderer.getTableCellRendererComponent(OutlineView.java:2258)
	at org.netbeans.swing.etable.ETableHeader$ETableHeaderRenderer.getTableCellRendererComponent(ETableHeader.java:129)
```
and there were no code changes in that area since quite a while.

my only explanation I can come up with right now is that:
https://github.com/apache/netbeans/blob/ac3936a4bdc7828721f45bba2feece190bc99a43/platform/openide.explorer/src/org/openide/explorer/view/OutlineView.java#L2196-L2200

is creating a cycle in some situations (fast scrolling?), when `createDefaultHeaderRenderer` is returning the wrong renderer?

**Comment by MKuettner** [comment]
The dialog is invoked from a NodeAction by calling `DialogDisplayer.getDefault().notify(..descriptor..)` so this should automatically run on EDT. I also added some `SwingUtilities.isEventDispatchThread()` check that are all pass.

This problem is also hard to debug since all the involved methods are called so many times without any problem.
I will try to extract some kind of test case.

**Comment by hclerx** [comment]
We had the same problem and we found it is caused by FlatLaf. We fixed it by patching FlatTableHeaderUI as follows in the paint() method:

		// temporary use own default renderer if necessary
		FlatTableCellHeaderRenderer tempRenderer = null;
		if ( !header.getDefaultRenderer().getClass().getName().equals("org.netbeans.swing.etable.ETableHeader$ETableHeaderRenderer") ) {
			// temporary use own default renderer
			tempRenderer = new FlatTableCellHeaderRenderer( header.getDefaultRenderer() );
			header.setDefaultRenderer( tempRenderer );
		}

		// paint header
		super.paint( g, c );

		// restore default renderer
		if( tempRenderer != null ) {
			tempRenderer.reset();
			header.setDefaultRenderer( tempRenderer.delegate );
		}

**Comment by mbien** [comment]
@hclerx interesting, thanks for posting your workaround!

cc @DevCharly

**Comment by DevCharly** [comment]
I've reworked FlatLaf's `FlatTableHeaderUI`.  (commit https://github.com/JFormDesigner/FlatLaf/commit/4df34b3f9d47539e48fde888f0b9b2a2dec9bc92)
It no longer temporary replaces cell renderer while painting header, which seems to cause the stack overflow in NB.

Please try latest FlatLaf `3.3-SNAPSHOT`: https://github.com/JFormDesigner/FlatLaf#snapshots

**Comment by MKuettner** [comment]
This seems to fix the problem. 
I'm not able to reproducte the StackOverFlowError anymore using FlatLaf `3.3-SNAPSHOT`.

It would be great to have this version available with one of the next Netbeans releases.
Thanks!

**Comment by oyarzun** [comment]
@DevCharly do you have an ETA for when FlatLAF v3.3 will be released or maybe a v3.2.6 bug fix release? It would be nice to get this fix in NetBeans 21.
